File: backend/app/models/body_composition.rb
class BodyComposition < ApplicationRecord
  validates :time, presence: true, uniqueness: true
  validates :weight, :body_fat, :body_fat_mass, :visceral_fat_level, :basal_metabolism,
            :muscle_mass, :bone_mass, :body_water_percentage,
            numericality: { greater_than: 0 }, allow_nil: true
  validates :body_age, :muscle_quality_score_all,
            numericality: { greater_than: 0 }, allow_nil: true

  scope :ordered, -> { order(time: :desc) }
  scope :on_date, ->(date) { where(time: date.all_day) }

  def to_display_hash
    attributes.except('id', 'created_at', 'updated_at')
  end
end

File: frontend/app/lib/body-composition.ts
/**
 * BodyComposition型定義とAPI関数
 */

import { apiClient } from './api';

export interface BodyComposition {
  time: string;
  weight?: number;
  body_fat?: number;
  body_fat_mass?: number;
  visceral_fat_level?: number;
  basal_metabolism?: number;
  muscle_mass?: number;
  bone_mass?: number;
  body_age?: number;
  body_water_percentage?: number;
  muscle_quality_score_all?: number;
}

/**
 * 体組成データ一覧を取得
 */
export async function getBodyCompositions(): Promise<BodyComposition[]> {
  return apiClient.get<BodyComposition[]>('/body_compositions');
}